### ISSUE TITLE
Should inputs and outputs be tied to the transaction?

### ISSUE THREAD

**Issue by SebastianStehle** [LSViana/ydotnet#45]
Some inputs and outputs allocate memory, which has to be freed again. Afaik this does not happen automatically. Therefore every input needs to be released again after the transaction has been completed.

This causes 2 issues IMHO:

1. It is not so easy to understand when an input need to be released. Because typically, when you add something to a map or array you don't want the items to be released before the container.
2. The output is even more complicated because you want to have the CLR value longer, but not the output itself.
3. It is just easy to forget in general.
4. Sometimes there is no container. Lets say you get a `output.Collection` or `output.Array` where you have to loop over the outputs manually and dispose them.

So in general there are not hat much uses cases ,where an input and output needs to live outside the transaction.

What I recommend is the following:

1. Maintain a list of all allocated resources within an transaction and then release them when the transaction is disposed.
2. Try to refactor the output so that it resolves the CRL value in the constructor and releases the actual output pointer immediately. I thin there is no need to keep that in memory and the use cases where you need an output without the underlaying value should be very rate.

**Comment by SebastianStehle** [comment]
I tried to improve memory allocations but it is actually super complicated. The problem is that every object seems to live in a scope and it actually depends on the method call and therefore Dispose does not really help.

lets take array as an example:

* If you create an array from a doc, you should actually never really dispose it until the document is disposed. But once the parent scope (the document) is disposed, you cannot work with the array anymore, so we must prevent that with guarding every single method. Otherwise it would cause issues with memory access and your process can crash.

* If you resolve the array via an output, you cannot dispose it, because it is actually disposed when the output is released and this is very tricky, because it depends whether you use an event or not.

So my idea would be the following:

We create scopes and scoped objects. Each scoped object needs a reference to a scope and can also add resources to that, something like that:

```
class Scope : IDisposable
{
   public bool IsDisposed { get; private set; }
   
   internal void AddResource(IDisposable)
   {
   }
}

class ScopedObject
{
    protected ScopedObject(Scope scope)
    {
    }
    
    protected void ThrowIfDisposed() {
         if (scope.IsDisposed)
         {
         }
    }
}
```

Scopes could be: Doc, Transaction, Event.

**Comment by SebastianStehle** [comment]
I tried to improve memory allocations but it is actually super complicated. The problem is that every object seems to live in a scope and it actually depends on the method call and therefore Dispose does not really help.

lets take output as an example:

* If you get an output from an object the output can actually be destroyed with `youtput_destroy`. So we could implement a dispose method and a finalizer to destroy the output when it is not needed anymore.
* If the output is part on json array via `youtput_read_json_array` it cannot be destroyed because it is destroyed as part of `youtput_destroy` of the json array.
* If the output is from an map iterator, it is actually super complicated. The output belongs to the map entry and cannot be destroy independently. But when do you destroy the map entry? 
  * If you destroy the map entry immediately, you cannot access the output later. This is especially true if you use something like FirstOrDefault which then also disposes the iterator / enumerator before the output is actually used.

We could avoid the problem with the output by resolving all values directly, but this would hurt performance, especially for events.

So on idea would be to have scopes, which are> 

* Doc
* Transaction
* Event

Every object is a resource and has a reference to a scope. Once a scope is over it releases all children and makes them unusable. If you use any method a ObjectDisposedException is thrown.

**Comment by LSViana** [comment]
Hi, @SebastianStehle!

About this:

> Some inputs and outputs allocate memory, which has to be freed again. Afaik this does not happen automatically. Therefore every input needs to be released again after the transaction has been completed.

Yes, that's true. But I'm working on multiple improvements to the memory management of the library to free resources via finalizers.

Right now, every exposed resource that can be disposed implements `IDisposable`, but they don't implement finalizers to allow the GC to trigger the dispose procedure.

So, this means that today if the client of the library doesn't dispose the `Input` cells, for example, they'll have a memory leak. This should be fixed after #34 is finished.

**Comment by SebastianStehle** [comment]
I came to the same conclusion now with finalizers. In my own fork I am going for a simplified approach for now, that does not need dispose at all, because almost all memory can be released immediately. 

If you want to keep the output you need a reference to a resource owner to keep the resource owner alive, otherwise something like that could fail:

```
var x = map.Iterate().Where(x => x ...).FirstOrDefault();

// GC.Collect();
x.DoSomething(); // Might crash because the map entry has been released already.
```

**Comment by LSViana** [comment]
Hi, @SebastianStehle!

Same as other issues in the repo, I believe this one can be closed now that we merged #55, right?
